### PR TITLE
QF-2711 Adding a few failsafes around ListFilesView.get()

### PR DESCRIPTION
**Pull request by why-not-try-calmer**
_Source_:
https://app.clickup.com/t/2192114/QF-2711.

_Problem_:
The `download_project` step used in the `process_projectfile` workflow makes a call to `qfieldcloud_sdk` to the effect that the sdk will try and GET project files under a given project id. However, the attempt sometimes fails on "500: Internal Error", with the caller being offered little to no information about the error. 

_Proposal_:
This PR proposes to offer callers more information about such failures. It delineates all the I/O actions that might raise, collect the possible exceptions, and returns the first one raised as part of the response.

### CODE REVIEW

**Comment by duke-nyuki** [comment]
Task linked: [QF-2711 Investigate and specify API_INTERNAL_SERVER_ERROR](https://app.clickup.com/t/2192114/QF-2711)

**Comment by suricactus** [comment]
I think in this case we can already use `ProjectMixin`, which handles the situation with wrong project id?

File: docker-app/qfieldcloud/core/views/files_views.py [hunk 1]
```diff
@@ -38,9 +40,17 @@ class ListFilesView(views.APIView):
...
 
-        project = Project.objects.get(id=projectid)
-
-        bucket = utils.get_s3_bucket()
+        try:
+            project = Project.objects.get(id=projectid)
+            bucket = utils.get_s3_bucket()
+            assert hasattr(bucket, "object_versions")
+        except ObjectDoesNotExist:
+            return bad_request(
```
[the remark this answers is not in the document]
Comment (faebebin) [reply docker-app/qfieldcloud/core/views/files_views.py:48]: 404 sounds reaonable, just an inexistent resource?

**Review by faebebin — CHANGES_REQUESTED** [review]

File: docker-app/qfieldcloud/core/views/files_views.py [hunk 2]
```diff
@@ -37,10 +39,21 @@ class ListFilesView(views.APIView):
...
+            if not bucket.creation_date:
+                # Let DRF return 500 when bucket does not exist, since it needed but
+                # not what the client tried to get
+                return server_error(
+                    request=request,
+                    reason=f"Unable to fetch needed resource for {projectid}",
+                )
+        except ObjectDoesNotExist:
+            # map failure to get from db into failure to GET from API
+            raise NotFound(detail=projectid)
```
Comment (faebebin) [inline docker-app/qfieldcloud/core/views/files_views.py:54]: Does this return a 404?

**Review by faebebin — CHANGES_REQUESTED** [review]
Think looks very promising?

Maybe could you describe the error handling behaviour in a small test case in public/docker-app/qfieldcloud/core/tests/test_qgis_file.py ? 

e.g `returns 404 if project with requested id does not exist`

File: docker-app/qfieldcloud/core/views/files_views.py (hunk 2 above)
Comment (why-not-try-calmer) [reply docker-app/qfieldcloud/core/views/files_views.py:54]: yes

File: docker-app/qfieldcloud/core/tests/test_qgis_file.py [hunk 3]
```diff
@@ -56,6 +56,16 @@ def get_file_contents(self, project, filename, version=None):
...
 
+    def test_try_to_get_nonexistent_project(self):
+        empty_string = ""
+        nonexistent_id = "007"
+        with self.subTest(
+            "Ensure that '/api/v1/files' handles missing resources correctly."
+        ):
+            one = self.client.get(f"/api/v1/files/{empty_string}")
+            two = self.client.get(f"/api/v1/files/{nonexistent_id}")
+            assert all(r.status_code == 404 for r in (one, two))
```
Comment (faebebin) [inline docker-app/qfieldcloud/core/tests/test_qgis_file.py:67]: Just subjective little taste comment from myside: I would keep it simple such a test and rather just write `assertEqual(one.status_code, 404) etc` but is ok

**Review by faebebin — APPROVED** [review]
LGTM, thanks :)

File: docker-app/qfieldcloud/core/tests/test_qgis_file.py (hunk 3 above)
Comment (suricactus) [reply docker-app/qfieldcloud/core/tests/test_qgis_file.py:67]: Forget about it, I will handle it.

Comment (faebebin) [reply docker-app/qfieldcloud/core/tests/test_qgis_file.py:67]: Thanks for the correction

**Pull request merged**